apps/arweave/src/ar_pool.erl

```diff
@@ -0,0 +1,766 @@
+	gen_server:call(?MODULE, get_current_sessions, infinity).
+
+%% @doc Return a set of the most recent cached jobs.
+get_jobs(PrevOutput) ->
+	gen_server:call(?MODULE, {get_jobs, PrevOutput}, infinity).
+
+%% @doc Return the most recent cached VDF output. Return an empty binary if the
+%% cache is empty.
+get_latest_output() ->
+	gen_server:call(?MODULE, get_latest_output, infinity).
+
+%% @doc Cache the given jobs.
+cache_jobs(Jobs) ->
+	gen_server:cast(?MODULE, {cache_jobs, Jobs}).
+
+%% @doc Validate the given (partial) solution. If the solution is eligible for
+%% producing a block, produce and publish a block.
+process_partial_solution(Solution) ->
+	gen_server:call(?MODULE, {process_partial_solution, Solution}, infinity).
+
+%% @doc Send the partial solution to the pool.
+post_partial_solution(Solution) ->
+	gen_server:cast(?MODULE, {post_partial_solution, Solution}).
+
+%%%===================================================================
+%%% Generic server callbacks.
+%%%===================================================================
+
+init([]) ->
+	process_flag(trap_exit, true),
+	ok = ar_events:subscribe(solution),
+	{ok, #state{}}.
+
+handle_call(get_current_sessions, _From, State) ->
+	{reply, lists:sublist(State#state.session_keys, 2), State};
+
+handle_call({get_jobs, PrevOutput}, _From, State) ->
+	SessionKeys = State#state.session_keys,
+	JobCache = State#state.jobs_by_session_key,
+	{reply, get_jobs(PrevOutput, SessionKeys, JobCache), State};
+
+handle_call(get_latest_output, _From, State) ->
+	case State#state.session_keys of
+		[] ->
+			{reply, <<>>, State};
+		[Key | _] ->
+			{O, _SN, _U, _S, _Diff} = hd(maps:get(Key, State#state.jobs_by_session_key)),
+			{reply, O, State}
+	end;
+
+handle_call({process_partial_solution, Solution}, From, State) ->
+	#state{ request_pid_by_ref = Map } = State,
+	Ref = make_ref(),
+	case process_partial_solution(Solution, Ref) of
+		noreply ->
+			{noreply, State#state{ request_pid_by_ref = maps:put(Ref, From, Map) }};
+		Reply ->
+			{reply, Reply, State}
+	end;
+
+handle_call(Request, _From, State) ->
+	?LOG_WARNING([{event, unhandled_call}, {module, ?MODULE}, {request, Request}]),
+	{reply, ok, State}.
+
+handle_cast({cache_jobs, #jobs{ jobs = [] }}, State) ->
+	{noreply, State};
+handle_cast({cache_jobs, Jobs}, State) ->
+	#jobs{ jobs = JobList, partial_diff = Diff,
+			next_seed = NextSeed, seed = Seed,
+			interval_number = IntervalNumber,
+			next_vdf_difficulty = NextVDFDifficulty } = Jobs,
+	SessionKey = {NextSeed, IntervalNumber, NextVDFDifficulty},
+	SessionKeys = State#state.session_keys,
+	SessionKeys2 = [SessionKey | SessionKeys],
+	JobList = [{Job#job.output, Job#job.global_step_number,
+			Job#job.partition_upper_bound, Seed, Diff} || Job <- JobList],
+	PrevJobList = maps:get(SessionKey, State#state.jobs_by_session_key, []),
+	JobList2 = JobList ++ PrevJobList,
+	JobsBySessionKey = maps:put(SessionKey, JobList2, State#state.jobs_by_session_key),
```

Comment:
Accidental `JobList` reuse (left and right side) causes badmatch. JobList on left side should be renamed
```
JobList2 = [{Job#job.output, Job#job.global_step_number,
			Job#job.partition_upper_bound, Seed, Diff} || Job <- JobList],

PrevJobList = maps:get(SessionKey, State#state.jobs_by_session_key, []),
JobList3 = JobList2 ++ PrevJobList,
JobsBySessionKey = maps:put(SessionKey, JobList3, State#state.jobs_by_session_key),
```